### ISSUE TITLE
ActiveRecord::Persistence#upsert_all throws PG::CardinalityViolation: ERROR:  ON CONFLICT DO UPDATE command cannot affect row a second time

### ISSUE THREAD

**Issue by vishaltelangre** [rails/rails#35519]
### Steps to reproduce
1. Create `upsert_all_issue` database on PostgreSQL using `createdb upsert_all_issue` command.
2. Use the following script to reproduce the issue.

```rb

# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "upsert_all_issue", username: "postgres", password: "")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :articles, force: true do |t|
    t.string :title, null: false
    t.string :slug, null: false
    t.string :author, null: false

    t.index :slug, unique: true
  end
end

class Article < ActiveRecord::Base; end

class BugTest < Minitest::Test
  def test_upsert_all
    Article.upsert_all(
      [
        {title: "Article 1", slug: "article", author: "Jane"},
        {title: "Article 2", slug: "article", author: "John"}
      ],
      unique_by: { columns: %w[ slug ] }
    )

    assert_equal("Article 2", Article.first.title)
  end
end
```

### Expected behavior

As per https://github.com/rails/rails/pull/35077, `Article.upsert_all` should insert just only one article record with title `Article 2` instead of inserting two separate records since the `slug` attribute in both of the attribute hashes is same (or duplicate).

### Actual behavior

The upsert_all query

```rb
Article.upsert_all(
  [
    {title: "Article 1", slug: "article", author: "Jane"},
    {title: "Article 2", slug: "article", author: "John"}
  ], 
  unique_by: { columns: %w[ slug ] }
)
``` 
panics with the following exception.

```
  Bulk Insert (4.9ms)  INSERT INTO "articles"("title","slug","author") VALUES ('Article 1', 'article', 'Jane'), ('Article 2', 'article', 'John') ON CONFLICT ("slug") DO UPDATE SET "title"=excluded."title","author"=excluded."author" RETURNING "id"
Traceback (most recent call last):
        1: from (irb):1
ActiveRecord::StatementInvalid (PG::CardinalityViolation: ERROR:  ON CONFLICT DO UPDATE command cannot affect row a second time)
HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.
```

### System configuration

**Rails version**:
rails 6.0.0.beta2 (at master@a62683f)

**Ruby version**:
2.6.0

**PostgreSQL version**:
10.5

**Comment by vishaltelangre** [comment]
The same script (below one) with `SQLite3` works just fine. So it seems that the issue is only related to PostgreSQL.

```rb

# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: "upsert_all_issue")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :articles, force: true do |t|
    t.string :title, null: false
    t.string :slug, null: false
    t.string :author, null: false

    t.index :slug, unique: true
  end
end

class Article < ActiveRecord::Base; end

class BugTest < Minitest::Test
  def test_upsert_all
    Article.upsert_all(
      [
        {title: "Article 1", slug: "article", author: "Jane"},
        {title: "Article 2", slug: "article", author: "John"}
      ],
      unique_by: { columns: %w[ slug ] }
    )

    assert_equal("Article 2", Article.first.title)
    assert_equal(1, Article.count)
  end
end
```

##### Output of the above script

```
[...]

Finished in 0.012828s, 77.9545 runs/s, 155.9090 assertions/s.
1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by boblail** [comment]
Did you notice how it works on Postgres if you split up the commands?

```ruby
Article.insert(
  { title: "Article 1", slug: "article", author: "Jane" }
)
Article.upsert(
  { title: "Article 2", slug: "article", author: "John" },
  unique_by: { columns: %w[ slug ] }
)
```

Postgres deliberately takes the position that
1. It should not update a row twice in one command (i.e. inserting it _and_ updating it)
2. Users can and should de-dupe input to the INSERT command

[There's a good comment in Postgres' source](https://github.com/postgres/postgres/blob/master/src/backend/executor/nodeModifyTable.c#L1335-L1355) where the exception is raised, explaining the database's rationale.

**Comment by vishaltelangre** [comment]
@boblail Thank you for sharing the link to the Postgres' source. It is really helpful.

I reported this issue in the first place since I couldn't get the `#upsert_all` working by following the documentation of `ActiveRecord::Persistence#upsert_all` here - https://github.com/rails/rails/blob/4b68e6dbf40558e06a5ceb2b9e1642d798b188ad/activerecord/lib/active_record/persistence.rb#L243-L249. 

The specified example in the documentation for upserting two book values in a single query with duplicate `isbn` doesn't work. The script here (https://gist.github.com/vishaltelangre/ba7ad59509f1140631ca45e6d7d48595) uses the book example to reproduce the same issue.

Do you think that we should update the documentation accordingly so that people will not run into the same issue?

**Comment by boblail** [comment]
@vishaltelangre, ah! Thank you so much for pointing that out! 🤦‍♂️ I'll create a PR to adjust the docs.

**Comment by boblail** [comment]
For the record, I did the script with MySQL (can't use `unique_by` with MySQL, but it's not strictly necessary in this test):

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "mysql2"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "mysql2", database: "activerecord_unittest")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :articles, force: true do |t|
    t.string :title, null: false
    t.string :slug, null: false
    t.string :author, null: false

    t.index :slug, unique: true
  end
end

class Article < ActiveRecord::Base; end

class BugTest < Minitest::Test
  def test_upsert_all
    Article.upsert_all(
      [
        {title: "Article 1", slug: "article", author: "Jane"},
        {title: "Article 2", slug: "article", author: "John"}
      ]
    )

    assert_equal("Article 2", Article.first.title)
  end
end
```

And the test passed.

**Comment by boblail** [comment]
:thinking: So of the three, only Postgres seems to want to avoid this double-update scenario; but I _think_ it's OK for lower-level methods like this to introduces fewer layers of padding between you and your database — i.e. OK for Postgres users experience this exception.

I definitely agree that the examples in the docs should just work out of the box!

**Comment by kamipo** [comment]
the docs is fixed by #35531.
